test/unit/math/serializer.hpp

```diff
@@ -92,10 +93,30 @@ struct deserializer {
    * @param x pattern argument to determine result shape and size
    * @return deserialized value with shape and size matching argument
    */
-  template <typename U>
-  typename stan::math::promote_scalar_type<T, std::vector<U>>::type read(
-      const std::vector<U>& x) {
-    typename stan::math::promote_scalar_type<T, std::vector<U>>::type y;
+  template <typename U, require_std_vector_t<U>* = nullptr,
+            require_not_std_vector_st<is_complex, U>* = nullptr>
+  typename stan::math::promote_scalar_type<T, U>::type read(const U& x) {
+    typename stan::math::promote_scalar_type<T, U>::type y;
+    y.reserve(x.size());
+    for (size_t i = 0; i < x.size(); ++i)
+      y.push_back(read(x[i]));
+    return y;
```

Comment:
Why do we even need separate overloads for complex/non-complex? Code seems exactly the same.

Reply:
It's mostly so that the return type gets instantiated correctly. The current state of the AD testing framework has all of the templating referring to the type within the complex variable (```std::complex<T>```), so these templates need to do the same (even if the function calls themselves are near-identical). Otherwise I'd need to update the std::complex templating throughout the rest of the framework (so that ```T``` refers to ```std::complex<double>``` rather than just ```double```) to collapse these two definitions to one.

Reply:
I think second option would lead to much clearer code. Unless that is too much work for this PR?

Reply:
Yeah it would be a bit much just for the result to be collapsing these two definitions, the new definitions are doc-ed as being for complex inputs, so I figure that's explanatory enough